mpc/mpc-aio/src/protocol/garble/exec/zk/prover.rs

```diff
@@ -0,0 +1,264 @@
+            .await?;
+
+        Ok(Prover {
+            state: Validate {
+                prover,
+                input_state: self.state.input_state,
+            },
+            circ: self.circ,
+            channel: self.channel,
+            backend: self.backend,
+            label_receiver: self.label_receiver,
+        })
+    }
+}
+
+impl<B, LS> Prover<Validate, B, LS>
+where
+    B: Evaluator + Compressor + Validator + Send,
+    LS: ObliviousReceive<InputValue, ActiveEncodedInput> + ObliviousVerify<FullEncodedInput> + Send,
+{
+    /// Execute the final phase of the protocol. This proves the authenticity of the circuit output
+    /// to the Verifier without leaking any information about the Prover's inputs.
+    pub async fn prove(mut self) -> Result<(), GCError> {
+        // Receive circuit opening to follower's circuit
+        let msg = expect_msg_or_err!(
+            self.channel.next().await,
+            GarbleMessage::CircuitOpening,
+            GCError::Unexpected
+        )?;
+
+        // Check circuit opening and pull it out for async validation
+        let (opening, gc_cmp, prover) = self.state.prover.validate_external(msg.into())?;
+
+        // Reconstruct full input labels from opening
+        let input_labels = FullInputSet::from_decoding(
+            gc_cmp.input_labels().clone(),
+            opening.get_delta(),
+            opening.get_decoding().to_vec(),
+        )?;
+
+        // Concurrently validate OT and garbling
+        let gc_validate_fut = self.backend.validate_compressed(gc_cmp, opening);
+
+        // If we did not receive any inputs via OT, we can skip the OT validation
+        let labels_validate_fut = if self.state.input_state.ot_receive_inputs.is_empty() {
+            Box::pin(ready(Ok(())))
+        } else {
+            let Some(label_receiver) = self.label_receiver.take() else {
+                return Err(GCError::MissingOTReceiver);
+            };
+
+            let ot_received = self
+                .state
+                .input_state
+                .ot_receive_inputs
+                .iter()
+                .map(|input| {
+                    input_labels
+                        .get(input.index())
+                        .expect("Input id should be valid")
+                })
+                .cloned()
+                .collect::<Vec<_>>();
+
+            label_receiver.verify(ot_received)
+        };
+
+        let (gc_validate_result, labels_validate_result) =
+            futures::join!(gc_validate_fut, labels_validate_fut);
+
+        // Both of these must pass to continue
+        _ = gc_validate_result?;
+        _ = labels_validate_result?;
+
+        // Reveal output to follower
+        let (commit_opening, output) = prover.reveal();
+
+        self.channel
+            .feed(GarbleMessage::CommitmentOpening(commit_opening.into()))
+            .await?;
+
+        self.channel
+            .send(GarbleMessage::Output(output.into()))
```

Comment:
what's the purpose of this? these are the same labels as in CommitmentOpening above

Reply:
Do you plan it as a future optimization that the prover sends the actual output instead of output labels and then verifier reconstructs the commitment?